### PR TITLE
Replace Plutus Addresses by Cardano Addresses

### PR DESCRIPTION
**Pull request by berewt**
Still a WIP, but we're close.

Pre-submit checklist:
- Branch
    - [ ] Tests are provided (if possible)
    - [ ] Commit sequence broadly makes sense
    - [ ] Key commits have useful messages
    - [ ] Formatting, PNG optimization, etc. are updated
    - [ ] Important changes are reflected in changelog.d of the affected packages
- PR
    - [ ] Self-reviewed the diff
    - [ ] Useful pull request description
    - [ ] Reference the ADR in the PR and reference the PR in the ADR (if revelant)
    - [ ] Reviewer requested

### CODE REVIEW

File: plutus-contract/src/Plutus/Contract/StateMachine/OnChain.hs [hunk 1]
```diff
@@ -104,8 +106,8 @@ data StateMachineInstance s i = StateMachineInstance {
     typedValidator :: TypedValidator (StateMachine s i)
     }
 
-machineAddress :: StateMachineInstance s i -> Address
-machineAddress = validatorAddress . typedValidator
+machineAddress :: StateMachineInstance s i -> CardanoAddress
+machineAddress = validatorCardanoAddress testnet . typedValidator
```
Comment (koslambrou) [inline plutus-contract/src/Plutus/Contract/StateMachine/OnChain.hs:110]: Add a TODO comment here saying that anything written with the StateMachine won't be usable on mainnet.

File: plutus-contract/test/Spec/Contract/Tx/Constraints/MustReferenceOutput.hs [hunk 2]
```diff
@@ -399,7 +402,7 @@ mustReferenceOutputTxV2Contract = do
     submitTxConfirmed $ mkTx lookups1 tx1
 
     -- Trying to unlock the Ada in the script address
-    scriptUtxos <- utxosAt mustReferenceOutputV2ValidatorAddress
+    scriptUtxos <- utxosAt $ mustReferenceOutputV2ValidatorAddress
```
Comment (koslambrou) [inline plutus-contract/test/Spec/Contract/Tx/Constraints/MustReferenceOutput.hs:405]: Remove $

File: plutus-contract/test/Spec/Contract/Tx/Constraints/MustSpendPubKeyOutput.hs [hunk 3]
```diff
@@ -73,15 +74,16 @@ mustSpendPubKeyOutputContract = mustSpendPubKeyOutputContract' []
...
+    params <- getParams
     let lookups1 = Constraints.typedValidatorLookups typedValidator
         tx1 = Constraints.mustPayToTheScriptWithDatumInTx onChainTxOutRefs (Ada.lovelaceValueOf baseLovelaceLockedByScript)
             <> foldMap Constraints.mustBeSignedBy keys
     ledgerTx1 <- submitTxConstraintsWith lookups1 tx1
     awaitTxConfirmed $ Tx.getCardanoTxId ledgerTx1
 
-    pubKeyUtxos <- utxosAt $ Ledger.pubKeyHashAddress pkh Nothing
+    pubKeyUtxos <- utxosAt $ either (error "can't build address") id $ toCardanoAddressInEra (Ledger.pNetworkId params) $ Ledger.pubKeyHashAddress pkh Nothing
     logInfo @String $ "pubKeyUtxos:: " ++ show pubKeyUtxos -- remove
```
Comment (koslambrou) [inline plutus-contract/test/Spec/Contract/Tx/Constraints/MustSpendPubKeyOutput.hs:85]: bin

**Review by koslambrou — APPROVED** [review]

**Review by james-iohk — APPROVED** [review]

File: plutus-script-utils/src/Plutus/Script/Utils/Typed.hs [hunk 4]
```diff
@@ -87,6 +91,14 @@ validatorHash = tvValidatorHash
 validatorAddress :: TypedValidator a -> PV1.Address
 validatorAddress = PV1.scriptHashAddress . tvValidatorHash
 
+-- | The address of the validator.
+validatorCardanoAddress :: C.NetworkId -> TypedValidator a -> C.AddressInEra C.BabbageEra
+validatorCardanoAddress networkId tv = let
+  validator = tvValidator tv
```
Comment (sjoerdvisscher) [inline plutus-script-utils/src/Plutus/Script/Utils/Typed.hs:97]: Put `let` before `validator` and indent the cases.

File: plutus-ledger/src/Ledger/Tx/CardanoAPI/Internal.hs [hunk 5]
```diff
@@ -458,10 +458,11 @@ toCardanoTxOut networkId (PV2.TxOut addr value datum _rsHash) =
             <*> toCardanoTxOutDatum datum
             <*> pure C.ReferenceScriptNone -- Not possible from just a hash
 
+{-# DEPRECATED fromCardanoAddressInEra "we now use Cardano address internally, if you need a plutus address use 'Ldeger.Adrress.toPlutusAddress' "#-}
```
Comment (sjoerdvisscher) [inline plutus-ledger/src/Ledger/Tx/CardanoAPI/Internal.hs:461]: Some typos in `Ldeger.Adrress.toPlutusAddress` :)

**Review by sjoerdvisscher — APPROVED** [review]
Great! Found some things that will make an improvement to my Value PR.

**Pull request merged**